intf/intf.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package intf implements mechanisms to access underlying interfaces in
+// a manner that is agnostic to the underlying implementation. An implementation
+// is included that uses netlink to write to a Linux kernl implementation.
+// Additional implementations can be provided to allow writing to e.g. a
+// gRPCwire implementation.
+package intf
+
+import (
+	"fmt"
+	"net"
+)
+
+// Interface represents information corresponding to a single interface on the
+// underlying system.
+type Interface struct {
+	// Index is an integer index used to refer to the interface.
+	Index int
+	// Name is the name used to refer to the interface in the system.
+	Name string
+	// MAC is the MAC address of the interface.
+	MAC net.HardwareAddr
+}
+
+// ARPEntry is a representation of an ARP entry on the underlying system.
+type ARPEntry struct {
+	// IP is the IP address of the neighbour.
+	IP net.IP
+	// Interface is the name of the interface on which the ARP entry was learnt.
+	Interface string
+	// MAC is the MAC address of the neighbour.
+	MAC net.HardwareAddr
+}
+
+// ARPEvent is used to describe a particular event on the ARP table.
+type ARPEvent int64
+
+const (
+	// ARPUnknown is used when the event is not an add or delete.
+	ARPUnknown ARPEvent = iota
+	// ARPAdd indicates that the ARP neighbour was added.
+	ARPAdd
+	// ARPDelete inidcates that the ARP neighbour was deleted.
+	ARPDelete
+)
+
+// ARPUpdate is used to describe a change to the ARP table.
+type ARPUpdate struct {
+	// Type indicates whether the event was an add or delete.
+	Type ARPEvent
+	// Neigh is the neighbour that the change corresponds to
+	Neigh ARPEntry
+}
+
+// NetworkAccessor is an interface implemented by the underlying system access
+// interface. Through implementing the NetworkAccessor interface it is possible
+// to use the functions within this package whilst using a different underlying
+// implementation (e.g., netlink on Linux to access the kernel).
+type NetworkAccessor interface {
+	// Links returns a set of links that are present on the local system.
```

Comment:
outdated comments? seems you named the functions to be around Interfaces, and not links

Reply:
for my understanding, can you also share about the reasoning for the change? why interface-oriented vs link-oriented access?

Reply:
Per our discussion - we normalised around the "interface" rather than link because it's more common to think about it in these terms. `netlink` uses the "link" term.